### PR TITLE
Another stab at tempfile

### PR DESCRIPTION
**Pull request by alemigo**
Async tempfile implementation following the existing structure, with a bit extra for SpooledTemporaryFile where delegation is not required unless the in-memory stream is rolled to disk.

### CODE REVIEW

**Comment by shaqarava** [comment]
Can you please fix the conflict if you have time? We need a tempfile rolled in. It's been years.

**Comment by alemigo** [comment]
@Tinche - do you have interest to add this?  Can update the pull request and add testing if helpful

**Comment by alemigo** [comment]
> Can you please fix the conflict if you have time? We need a tempfile rolled in. It's been years since adding it was first talked about. Seems like this one should have been merged.

updated

**Comment by CMCDragonkai** [comment]
Any update on whether this fixes #20?

**Comment by alemigo** [comment]
you can try https://github.com/alemigo/asynctempfile in the meantime

**Comment by CMCDragonkai** [comment]
I think this PR can be squashed to be ready for merge.

**Comment by wagnerflo** [comment]
Can a maintainer please confirm, that this will be merged so I can start using @alemigo's branch now and then switch back to a aiofiles release later?

**Comment by MajesticMullet** [comment]
Will this be merged soon?

**Comment by JavierMartinezAlc** [comment]
It will be really helpful if this is released

**Comment by alemigo** [comment]
https://github.com/alemigo/asynctempfile is the same code for tempfile support as this PR

**Comment by Tinche** [comment]
Alright, since there's interest I can merge this in. But it needs tests first.

**Comment by alemigo** [comment]
> Alright, since there's interest I can merge this in. But it needs tests first.

Ok I will add

**Comment by Tinche** [comment]
> > Alright, since there's interest I can merge this in. But it needs tests first.
> 
> Ok I will add

Excellent, thanks. Also some documentation on how to use it would be cool.

File: aiofiles/tempfile/__init__.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
+""" Async implementation of the tempfile module"""
+
+__version__ = '0.1.1'
```
Comment (Tinche) [inline aiofiles/tempfile/__init__.py:3]: Don't think we need this

File: aiofiles/tempfile/__init__.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+def TemporaryDirectory(loop=None, executor=None):
+    """Async open a temporary directory"""
+    return AiofilesContextManagerTempDir(_temporary_directory(loop=loop,
+                                                       executor=executor))
+
+
+# =========================================================
+# Internal coroutines to open new temp files/directories
+# =========================================================
+@coroutine
```
Comment (Tinche) [inline aiofiles/tempfile/__init__.py:72]: Can we switch to async/await for all new code? We support 3.5+ which has it.

**Review by Tinche — CHANGES_REQUESTED** [review]
Looks good, left a couple of comments. Also the PR won't merge cleanly for me, might have to rebase on master. Thanks!

**Comment by alemigo** [comment]
> Looks good, left a couple of comments. Also the PR won't merge cleanly for me, might have to rebase on master. Thanks!

Thx - made the changes

**Review by Tinche — APPROVED** [review]

**Comment by Tinche** [comment]
Looks good, thanks a lot!

**Pull request merged**